### DISCUSSION TITLE
Block layout: Gutenberg core blocks vs custom CSS

### DISCUSSION THREAD

**Discussion by danieldudzic** [woocommerce/woocommerce#44962] [WooCommerce, Blocks, and Site Editing]
I'm hoping to kick off a discussion about our "layout strategy" when building new blocks.

How much priority should Gutenberg core blocks have when it comes to replicating designed mockups?

At which point does the core-block-powered structure becomes overcomplicated, the cons outweigh the benefits and we should fallback on a custom CSS solution?

Can we discuss and agree upon some "rules of thumb" when it comes to making decision in this area?

<hr />

### Case studies

I'd like to bring up two new **Product Gallery** (inner) blocks as examples in this topic. The **Thumbnails** block has been worked on by me, and the **Next/Previous** buttons by @gigitux so I'm sure he'll have a bit more context and input regarding his block.

### 1. Product Gallery Thumbnails
![xOnjRFuOoH](https://github.com/woocommerce/woocommerce-blocks/assets/905781/33097810-554d-4f97-b21a-297d4a2e33ce)

After initially, trying things out with CSS and flexbox, @gigitux suggested giving the **Group** block a try. I have looked into it, and it turned out I could achieve the desired layout with the following configuration:

|Thumbnails on the left|Thumbnails on the bottom|Thumbnails on the right|
|--|--|--|
| <img width="647" alt="single_product_e280b9_template_part_e280b9_editor_e280b9_productgallery_e28094_wordpress-2" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/5c0372de-6355-4b4d-9fb1-33519982cb08">|<img width="650" alt="single_product_e280b9_template_part_e280b9_editor_e280b9_productgallery_e28094_wordpress-3" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/7f5032fc-11cc-4b84-a324-71e20d600750">|<img width="648" alt="cursor_i_single_product_e280b9_template_part_e280b9_editor_e280b9_productgallery_e28094_wordpress" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/45453238-94fe-4bf2-a0bb-e371dea65e68">|

I was then being able to add block settings manipulating the type of the **Group** block by updating the blocks attributes.

The downside of this, is that if for some reason there will be another **Group** block added to the **Product Gallery** it will also get updated.

The above solution adds a **Row/Stack** wrapper to the **Large Image** and **Thumbnails** blocks. This is of course reflected in the block **List View** and **Breadcrumbs**. I believe this is completely reasonable and doesn't "interrupt" the user experience too much. What do you think?

### 2. Product Gallery Next/Previous Buttons

The second case, is IMO a more interesting one.

@gigitux proposed the following block structure. Quoting Luigi:

- *The double **Stack** block is necessary to put the buttons at the bottom of the image.*

And he have argued:

- *This way merchants and developers will be able to place other items on the image (like the **Sale** badge).*
- *Have the freedom to change the position of the buttons easily.*

![image](https://github.com/woocommerce/woocommerce-blocks/assets/905781/77c83d43-2f69-4376-b7fd-cc93673c8083)

There's been some pushback to this in the related Slack discussion: p1689147821346709-slack-C02FL3X7KR6

The overcomplication of the block structure has been pointed out as a major downside, and CSS was recommended as the preferred solution.

<hr />

### Pros and cons of both solutions

Although I feel that nowadays there is an broad consensus and preference towards core blocks over CSS, I think it would be useful to spell it all out. This is a quick list that comes to my mind, but please feel free to point out anything I have missed.

#### Pros
- Generally future proof
- Reliance on core for maintenance
- Compatibility with Global Styles
- Fewer conflicts/anti-fragile and less likely to break
- Automatic upstream updates/upgrades

#### Cons

- Can become overly complex with nested blocks
- More difficult for beginners to understand and navigate

#### Pros
- Clean block structure ( List View and Breadcrumbs )
- Not limited by Gutenberg API and block styles
- Easier to understand layout and styles

#### Cons

- Great potential for conflicts
- No support for global styles
- Additional maintenance/testing/debugging
- Less code standardization

<hr />

Do we have any data on how users interact with blocks? The assumption I'd make is, that the majority of users utilize Gutenberg block controls and settings to interface with them. Given that fact, I do not think a complicated block/markup structure will be too bad of a disadvantage.

I believe the block structure issue, with cluttering of the markup, List View and Breadcrumbs is inherently a **Gutenberg Core** issue. Which should not stop us from using core blocks, but rather we should advocate for better core solution, and in the future when/if it ever arrives, we can easily migrate/benefit from it.

I believe that at the end of the day, better compatibility with the core will benefit our users.

Any feedback would be greatly appreciated 🙏

**Comment by gigitux** [comment]
Thanks for opening this discussion, @danieldudzic!

Regarding the case `Product Gallery Next/Previous Buttons`, I found two possible approaches to avoid the use of custom CSS:

### Using two stack blocks

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/78b89999-09e1-4fde-aabc-cb4cf0581f4e)

Currently, this is the only way to align a block at the bottom of its container. 
https://github.com/WordPress/gutenberg/discussions/48659

### Using Margin Bottom

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/6f6c0c81-6ebb-459c-94f5-d1c5cb84aedf)

The margin-bottom option isn't the best approach, but the merchant can easily customize the styles via the Global Styes settings. The same is true for theme developers. Via `theme.json`, they can put the button in any position in the container.

**Comment by nerrad** [comment]
Have you investigated utilizing the [block supports for `layout`](https://make.wordpress.org/core/2023/07/14/layout-updates-in-the-editor-for-wordpress-6-3/) directly in our custom block (instead of wrapping with a WP core block)? Even though it was stabilized to "layout" for WP 6.3 (thus requiring that as the minimum version), presumably we might be able to use the previous `__experimental` properties to support earlier WP versions too?

**Reply by danieldudzic** [reply]
Thank you for sharing, I was not aware of `layout`.

I have looked into it, but there's no direct way to modify `supports` dynamically based on block settings (at least I didn't find a way). It seems that it's a really good solution if you don't need to alter between different layout modes based on (block) settings.

If it would be possible to hook that value into block settings, it would resolve all issues with the overcomplicated structure (due to wrappers).

**Reply by nerrad** [reply]
What did you try?

**Reply by nerrad** [reply]
For instance group, row, and stack have transitions in the block settings:

<img width="1206" alt="CleanShot 2023-07-25 at 11 12 42@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/1429108/a8b2c0d2-dc5a-4376-96c6-5cf978792869">

But it looks like you've already tried implementing that?

In [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/9398), I was also able to implement directional layout options via the [supports.layout properties](https://github.com/woocommerce/woocommerce-blocks/pull/9398/files#diff-1a7b9fef8fd50a1e325296cce9617e3ae39d4cd38e7d67aaf90280d16cfe9cc8R10-R15).

Given the list of pros and cons, it appears to me that we'd want to implement a solution that mirrors the _behaviour_ of the core layout blocks but without the complexity of the nested blocks in the UI?

**Reply by danieldudzic** [reply]
I am able to update the `Group` block with a new layout value. That works great (as per my gif screenshot in the thread).

```
	const block = select( 'core/block-editor' ).getBlock( clientId );
	block?.innerBlocks.forEach( ( innerBlock ) => {
		if ( innerBlock.name === 'core/group' ) {
			const updatedBlock = {
				...innerBlock,
				attributes: {
					...innerBlock.attributes,
					layout: setGroupAttributes( attributes.thumbnailsPosition ),
				},
			};
			dispatch( 'core/block-editor' ).updateBlock(
				innerBlock.clientId,
				updatedBlock
			);
		}
	} );
```

I tried earlier today adding `layout` options via `block.json`:

```
{
	"name": "woocommerce/product-gallery",
	"title": "Product Gallery",
	"supports": {
		"__experimentalLayout": {
			"default": {
				"type": "flex",
				"flexWrap": "nowrap"
			},
			"allowEditing": false
		}
	},
}
```

I have tried both `__experimentalLayout` and `layout` and I was **successfully** able to enable the `flex` layout in the block.

But now it gets tricky. In `edit.tsx` I can now get the `layout` value via [getBlockSupport](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-blocks/#getblocksupport) but I wasn't able to find a good way to update it. AFAIK `setBlockSupport` does not exist. And when having a block setting changing the layout I need to be able to do it.
